### PR TITLE
_unitary_() init for qasm_output.py

### PR DESCRIPTION
**Pull request by vtomole**
Fixes https://github.com/quantumlib/Cirq/issues/1139

@Strilanc I don't think my implementation of `_unitary_` for QasmTwoQubitGate is correct as I tried to implement a test for CNOT, but it's not passing. Used: https://arxiv.org/pdf/quant-ph/0507171.pdf as a reference. Is it the test or the implementation?

### CODE REVIEW

File: cirq/circuits/qasm_output.py [hunk 1]
```diff
@@ -52,14 +53,24 @@ def _qasm_(self,
...
+            'u3({0:half_turns},{1:half_turns},{2:half_turns}) {3};\n',
+            self.theta, self.phi, self.lmda, qubits[0])
 
     def __repr__(self) -> str:
         return 'cirq.QasmUGate({}, {}, {})'.format(self.lmda,
                                                    self.theta,
                                                    self.phi)
 
+    def _unitary_(self) -> np.ndarray:
+        return np.array([[np.exp(-1j * (self.phi + self.lmda)/2) \
```
Comment (vtomole) [inline cirq/circuits/qasm_output.py]: Need space between `/`.

File: cirq/circuits/qasm_output.py [hunk 2]
```diff
@@ -52,14 +53,24 @@ def _qasm_(self,
...
+                          * np.cos(self.theta/2),
```
Comment (Strilanc) [inline cirq/circuits/qasm_output.py]: Simplify this by assigning variables for common expressions beforehand.

Alternatively, you may be able to simply take the matrix product of `Rz(a * pi) Rx(b * pi) Rz(c * pi)`.

File: cirq/circuits/qasm_output.py [hunk 3]
```diff
@@ -127,8 +138,24 @@ def _decompose_(self, qubits: Sequence[ops.QubitId]) -> ops.OP_TREE:
 
     def __repr__(self) -> str:
         return 'cirq.QasmTwoQubitGate({}, {}, {}, {}, {}, {}, {})'.format(
-                self.before0, self.before1, self.x, self.y, self.z,
-                self.after0, self.after1)
+            self.before0, self.before1, self.x, self.y, self.z,
+            self.after0, self.after1)
+
+    def _unitary_(self) -> np.ndarray:
```
Comment (Strilanc) [inline cirq/circuits/qasm_output.py]: This one got caught in the crossfire of https://github.com/quantumlib/Cirq/pull/1154 😨

File: cirq/circuits/qasm_output.py [hunk 4]
```diff
@@ -52,14 +53,20 @@ def _qasm_(self,
...
+            'u3({0:half_turns},{1:half_turns},{2:half_turns}) {3};\n',
+            self.theta, self.phi, self.lmda, qubits[0])
 
     def __repr__(self) -> str:
         return 'cirq.QasmUGate({}, {}, {})'.format(self.lmda,
                                                    self.theta,
                                                    self.phi)
 
+    def _unitary_(self) -> np.ndarray:
+        rz_phi_matrix = cirq.unitary(cirq.Rz(self.phi))
```
Comment (Strilanc) [inline cirq/circuits/qasm_output.py]: Have to divide angles by pi to convert from half-turns to radians.

**Comment by Strilanc** [comment]
I'm not sure what mypy is complaining about. It seems to think `cirq.unitary` is a module...?

**Comment by vtomole** [comment]
Strange. The test is not passing either.

**Comment by Strilanc** [comment]
Obsoleted by https://github.com/quantumlib/Cirq/pull/1402 (which just fixes the travis failures in this one; I can't push to your fork)

**Pull request closed without merging**